### PR TITLE
[SPARK-20483] Mesos Coarse mode may starve other Mesos frameworks

### PR DESCRIPTION
**Pull request by dgshep**
## What changes were proposed in this pull request?

Set maxCores to be a multiple of the smallest executor that can be launched. This ensures that we correctly detect the condition where no more executors will be launched when spark.cores.max is not a multiple of spark.executor.cores

## How was this patch tested?

This was manually tested with other sample frameworks measuring their incoming offers to determine if starvation would occur.

@dbtsai @mgummelt

### CODE REVIEW

**Comment by dbtsai** [comment]
Jenkins, add to whitelist.

**Comment by dbtsai** [comment]
Jenkins, ok to test.

**Comment by dbtsai** [comment]
LGTM. Merged into master and branch 2.2. Thanks.

**Comment by dbtsai** [comment]
Tests are added in a followup PR. https://github.com/apache/spark/pull/17788

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala [hunk 1]
```diff
@@ -60,8 +60,16 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
...
 
+  private val executorCoresOption = conf.getOption("spark.executor.cores").map(_.toInt)
+
+  private val minCoresPerExecutor = executorCoresOption.getOrElse(1)
+
   // Maximum number of cores to acquire
-  private val maxCores = maxCoresOption.getOrElse(Int.MaxValue)
+  private val maxCores = {
+    val cores = maxCoresOption.getOrElse(Int.MaxValue)
+    // Set maxCores to a multiple of smallest executor we can launch
```
Comment (dbtsai) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala]: Should we have a log here to let users know?

**Comment by mgummelt** [comment]
@dbtsai Can we please wait to get a LGTM from one of the active Mesos contributers (@skonto, @tnachen, myself, etc.) before merging Mesos code?

I would have rather this be solved in the offer rejection code rather than changing what `maxCores` means.

**Comment by mridulm** [comment]
I agree with @mgummelt, maxCores should have reflected user config option.

**Comment by dgshep** [comment]
Fair point. This felt like a succinct way to handle this corner case, but if it makes sense to harden the offer refusal code instead, I can update.

**Comment by dbtsai** [comment]
@mgummelt We tested this in our production env, and it solves our issue. Since it seems to be a trivial change, I made my judgement. Gonna wait for more feedback.

**Comment by dbtsai** [comment]
@mgummelt We tested this in our production env, and it solves our issue. Since it seems to be a trivial change, I made my judgement. Gonna wait for more feedback. Thanks.

**Pull request closed without merging**